core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java

```diff
@@ -0,0 +1,453 @@
+    }
+    subchannels.clear();
+    // NB(lukaszx0) Whether we should propagate the error unconditionally is arguable. It's fine
+    // for time being.
+    updateBalancingState(TRANSIENT_FAILURE, new Picker(PickResult.withError(error)));
+  }
+
+  void processSubchannelState(Subchannel subchannel, ConnectivityStateInfo stateInfo) {
+    ConnectivityState newState = stateInfo.getState();
+    // Shutdown channels/previously relevant subchannels can still callback with state updates.
+    // To prevent pickers from returning these obselete subchannels, this logic
+    // is included to check if the current list of active subchannels includes this subchannel.
+    if (!subchannel.equals(subchannels.get(subchannel.getAddresses().getAddresses().get(0)))) {
+      return;
+    }
+    if (newState == SHUTDOWN) {
+      return;
+    }
+    if (newState == IDLE) {
+      helper.refreshNameResolution();
+    }
+    // If we are transitioning from a TRANSIENT_FAILURE to CONNECTING or IDLE we ignore this state
+    // transition and still keep the LB in TRANSIENT_FAILURE state. This is referred to as "sticky
+    // transient failure". Only a subchannel state change to READY will get the LB out of
+    // TRANSIENT_FAILURE. If the state is IDLE we additionally request a new connection so that we
+    // keep retrying for a connection.
+
+    // With the new pick first implementation, individual subchannels will have their own backoff
+    // on a per-address basis. Thus, iterative requests for connections will not be requested
+    // once the first pass through is complete.
+    // However, every time there is an address update, we will perform a pass through for the new
+    // addresses in the updated list.
+    if (currentState == TRANSIENT_FAILURE) {
+      if (newState == CONNECTING) {
+        // each subchannel is responsible for its own backoff
+        return;
+      } else if (newState == IDLE) {
+        requestConnection();
+      }
+    }
+
+    states.put(subchannel, newState);
+    switch (newState) {
+      case IDLE:
+        // Shutdown when ready: connect from beginning when prompted
+        addressIndex.reset();
+        updateBalancingState(IDLE, new RequestConnectionPicker(this));;
+        break;
+      case CONNECTING:
+        // It's safe to use RequestConnectionPicker here, so when coming from IDLE we could leave
+        // the current picker in-place. But ignoring the potential optimization is simpler.
+        updateBalancingState(CONNECTING, new Picker(PickResult.withNoResult()));
+        break;
+      case READY:
+        updateBalancingState(READY, new Picker(PickResult.withSubchannel(subchannel)));
+        shutdownRemaining(subchannel.getAddresses().getAddresses().get(0), subchannel);
+        break;
+      case TRANSIENT_FAILURE:
+        // If we are looking at current channel, request a connection if possible
+        if (addressIndex.isValid() && subchannels.get(addressIndex.getCurrentAddress()) != null
```

Comment:
this probably can be simplified to `addressIndex.isValid && subchannel == subchannels.get(address)`

Reply:
Is there ever a case where `==` isn't enough?

Reply:
The null check is not necessary because `subchannel == null` return false. The equals method and `==` should be identical here. But that's minor.

Reply:
Should I replace all `equals()` with `==` to avoid any possible null pointers?